### ISSUE TITLE
Issue with highly-codependent variables with data generated by simple linear equation

### ISSUE THREAD

**Issue by parrt** [shap/shap#967]
howdy!  I have a simple linear equation that generates data but with a wrinkle; I've copied x1 into x3 and added some zero-centered low-amplitude noise to make them codependent but non-identical:

```
y = x_1 + x_2 + x_3 + 100 where x_3 = x_1 + noise, x_1, x_2 ~ U(0,10)
```

Ignoring x3 copy thing for the moment, we'd expect partial derivatives of y w.r.t. xi to be 1.0.  OLS indeed shows coefficients of 1.0 (of course with high variance on the betas).  SHAP `LinearExplainer` gives mean(abs(shap)) of [2.57113498 2.46506859 2.56234981], so equal weight to each coefficient.  By the way, should these be 1.0 also?

Now, x1 and x3 are not independent but I argue we should be able to tease apart x3's contribution as OLS and OLS shap are doing. Imagine we have enough data so we can find pairs of records with equal x1 and x3 values; they only differ in x3 due to noise:

<img width="211" alt="Screen Shot 2019-12-14 at 10 50 20 AM" src="https://user-images.githubusercontent.com/178777/71312166-9f438100-23dc-11ea-85dc-00e8be777170.png">

Seems like the change in y due solely to x3 (rise/run) would be:

```
(22+e2) - (22+e1)
------------------------
(9+e2) - (9+e1)
```

which reduces to 

```
e2 - e1
---------
e2 - e1
```

or 1.0 for any e1, e2 noise and any pair of stratified records.

Turning to a random forest model, we see x1 and x3 contributions bounce all over the place. As you explained in email, an RF has equally likely chances of using x1 or x3 when building decision trees so we would expect high variance between x1 and x3 contribution. However, wouldn't the expectation of x1 and x3 be 1.0 (or some fixed value...can't remember), just with high variance?  I ran a long experiment (50 trials) and they do seem to converge to the expectation.

For sure this example also highlights how one can get very different answers depending on the chosen model.

I enclose self-contained code to show OLS vs RF results.

Thanks for your attention!

```python
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm

import shap

import numpy as np
import pandas as pd

def synthetic_poly_dup_data(n):
    p = 3 # x1, x2, x3
    df = pd.DataFrame()
    for i in range(p):
        df[f'x{i + 1}'] = np.random.random_sample(size=n) * 10
    # copy x1 into x3 with zero-centered noise
    df['x3'] = df['x1'] + np.random.random_sample(size=n)-0.5
    yintercept = 100
    df['y'] = np.sum(df, axis=1) + yintercept
    terms = [f"x_{i+1}" for i in range(p)] + [f"{yintercept:.0f}"]
    eqn = "y = " + ' + '.join(terms) + " where x_3 = x_1 + noise, x_1, x_2 ~ U(0,10)"
    return df, eqn

df, eqn = synthetic_poly_dup_data(1000)
X = df.drop('y', axis=1)
y = df['y']

print(eqn)

# Use OLS to get coeff and then LinearExplainer

lm = LinearRegression()
lm.fit(X,y)
print("OLS coeff", lm.coef_)
y_pred = lm.predict(X)
print(f"OLS Training MSE {np.mean((y - y_pred) ** 2):.5f}")

explainer = shap.LinearExplainer(lm, X, feature_dependence='independent')
shap_values = explainer.shap_values(X)
shapimp = np.mean(np.abs(shap_values), axis=0)
print("OLS SHAP importances", shapimp)

beta_stderr = sm.OLS(y, X).fit().bse
print(f"linear model coefficient stderr {beta_stderr.values}\n(high variance on x1 and x3 as expected)")

# Try now with RF model and TreeExplainer

for i in range(5): # try 5 times to see variance in x_1, x_3
    print(f"Trial {i+1}")
    rf = RandomForestRegressor(n_estimators=20)
    rf.fit(X, y)
    print(f"RF Training MSE {np.mean((y - rf.predict(X)) ** 2):.5f}")
    explainer = shap.TreeExplainer(rf, data=X, feature_perturbation='interventional')
    shap_values = explainer.shap_values(X)
    shapimp = np.mean(np.abs(shap_values), axis=0)
    print("RF SHAP importances", shapimp)
```

OUTPUT:

```
y = x_1 + x_2 + x_3 + 100 where x_3 = x_1 + noise, x_1, x_2 ~ U(0,10)
OLS coeff [1. 1. 1.]
OLS Training MSE 0.00000
OLS SHAP importances [2.57113498 2.46506859 2.56234981]
linear model coefficient stderr [4.10717122 0.31065979 4.1116481 ]
(high variance on x1 and x3 as expected)
Trial 1
RF Training MSE 0.01661
RF SHAP importances [2.78986432 2.45207256 2.35204797]
Trial 2
RF Training MSE 0.01566
RF SHAP importances [2.32197286 2.42069104 2.81279345]
Trial 3
RF Training MSE 0.01574
RF SHAP importances [2.69656699 2.40900263 2.44039671]
Trial 4
RF Training MSE 0.01584
RF SHAP importances [2.95876692 2.42191879 2.18118951]
Trial 5
RF Training MSE 0.01641
RF SHAP importances [2.48588282 2.44207147 2.64593201]
```

**Comment by slundberg** [comment]
Thanks for the great example!

First, let's explain why mean(abs(shap_value)) is about 2.5 for each of the features (ignoring redundancy and assuming the linear model betas are 1.0). SHAP values essentially explain the "evidence" that takes you from your prior expectation of a model's output over some distribution to the current model's output. Since the model is the sum of three features (and we are ignoring dependence between them) we can consider each of those independently. The mean of a feature uniformly distributed between 0 and 10 is 5, so that is our prior expectation of the model output for just that feature (given we are doing least squares regression). The difference between 5 and the current output is the impact of the evidence provided by that feature. On average U(0,10) - 5 is 2.5.

This can be visualized with `shap.waterfall_plot(explainer.expected_value, shap_values[0], X.iloc[0,:])`:
![image](https://user-images.githubusercontent.com/3740613/71531618-52601e80-28a4-11ea-9b4f-8f25931a8e1b.png)

Note that because we are doing least squares regression the units on the x-axis can be interpreted as a log of our (normal) predictive probability distribution (i.e. units of information) so the length of each bar is the impact of the evidence the model extracts from each feature's value.

You will note that (x1+x3)/2 is close to 2.5 even in the RF case. Which means the RF has not evenly balanced the credit between x1 and x3 since either could be used for the other. Note also that the x2 importance is not exactly 2.5 which would suggest to me that the RF model is probably flattening out the true linear relationship a bit at the tails. Using `shap.dependence_plot(1, shap_values, X)` shows this is indeed the case:

![image](https://user-images.githubusercontent.com/3740613/71531978-25147000-28a6-11ea-84fb-8239fa400320.png)
